dd-list.1
.\" Copyright 2005 Lars Wirzenius
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH DD\-LIST 1 2005-06-28 "Debian"
.\" --------------------------------------------------------------------
.SH NAME
dd\-list \- nicely list .deb packages and their maintainers
.\" --------------------------------------------------------------------
.SH SYNOPSIS
.BR dd\-list " [" \-hiV "] [" \-\-help "] [" \-\-stdin "] ["
.BR \-\-dctrl "] [" \-\-version "] [" package "...]"
.\" --------------------------------------------------------------------
.SH DESCRIPTION
.B dd\-list
produces nicely formatted lists of Debian (.deb) packages and their
maintainers.
.PP
Input is a list of source or binary package names on the command line
(or the standard input if 
.B --stdin 
is given). 
Output is a list of the following format, where package names are source
packages:
.PP
.nf
.RS
J. Random Developer <jrandom@debian.org>
.RS
j-random-package
j-random-other
.RE
.PP
Diana Hacker <diana@example.org>
.RS
fun-package
more-fun-package
.RE
.RE
.fi
.PP
This is useful when you want, for example, to produce a list of packages
that need to attention from their maintainers, e.g., to be rebuilt when
a library version transition happens.
.\" --------------------------------------------------------------------
.SH OPTIONS
.TP
.BR \-h , \-\-help
Print brief help message.
.TP
.BR \-i , \-\-stdin
Read package names from the standard input, instead of taking them
from the command line. Package names are whitespace delimited.
.TP
.BR \-d , \-\-dctrl
Read package information from standard input in the format of a Debian
package control file. This includes the status file, or output of
apt-cache. This is the fastest way to use dd-list, as it uses the
maintainer information from the input instead of looking up the maintainer
of each listed package.
.TP
.BR \-V , \-\-version
Print the version.
.\" --------------------------------------------------------------------
.SH AUTHOR
Lars Wirzenius <liw@iki.fi>
.P
Joey Hess <joeyh@debian.org>


dd-list.pl
#!/usr/bin/perl
#
# dd-list: Generate a list of maintainers of packages.
#
# Written by Joey Hess <joeyh@debian.org>
# Based on a python implementation by Lars Wirzenius.
# Copyright 2005 Lars Wirzenius, Joey Hess
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

use warnings;
use strict;
use Getopt::Long;

my $version='###VERSION###';

sub get_developer_given_package {
	my $package_name=shift;
	
	my $developer;
	my $source_name;
	open (F, "apt-cache showsrc '$package_name' |");
	while (<F>) {
		chomp;
		if (/^Maintainer: (.*)/) {
			$developer=$1;
		}
		elsif (/^Package: (.*)/) {
			$source_name=$1;
		}
	}
	close F;
	return ($developer, $source_name);
}

sub parse_developer {
	my $developer=shift;

	my ($name, $domain)=$developer=~/^(.*)\s+<.*@(.*)>\s*$/i;
	if (defined $domain && $domain ne 'lists.debian.org') {
		return join " ", reverse split " ", $name;
	}
	elsif (defined $name) {
		return $name;
	}
	else {
		return $developer;
	}
}

sub sort_developers {
	sort { parse_developer($a) cmp parse_developer($b) } @_;
}

sub help {
	print <<"EOF"
Usage: dd-list [-hiV] [--help] [--stdin] [--dctrl] [--version] [package ...]

    -h, --help
        Print this help text.
        
    -i, --stdin
        Read package names from the standard input.

    -d, --dctrl
        Read Debian control data from standard input.
       
    -V, --version
        Print version (it's $version by the way).
EOF
}

my $use_stdin=0;
my $use_dctrl=0;
if (! GetOptions(
	"help" => sub { help(); exit },
	"stdin|i" => \$use_stdin,
	"dctrl|d" => \$use_dctrl,
	"version" => sub { print "dd-list version $version\n" })) {
	exit(1);
}

my %dict;
my $errors=0;

if ($use_dctrl) {
	$/="\n\n";
	while (<>) {
		my ($package, $maintainer);

		if (/^Package:\s+(.*)$/m) {
			$package=$1;
		}
		if (/^Source:\s+(.*)$/m) {
			$package=$1;
		}
		if (/^Maintainer:\s+(.*)$/m) {
			$maintainer=$1;
		}

		if (defined $maintainer && defined $package) {
			push @{$dict{$maintainer}}, $package;
		}
		else {
			print STDERR "E: parse error in stanza $.\n";
			$errors=1;
		}
	}
}
else {
	my @package_names;
	if ($use_stdin) {
		while (<>) {
			chomp;
			s/^\s+//;
			s/\s+$//;
			push @package_names, $_;
		}
	}
	else {
		@package_names=@ARGV;
	}

	foreach my $package_name (@package_names) {
		my ($developer, $source_name)=get_developer_given_package($package_name);
		if (defined $developer) {
			push @{$dict{$developer}}, $source_name;
		}
		else {
			print STDERR "E: Unknown package: $package_name\n";
			$errors=1;
		}
	}
}
	
foreach my $developer (sort_developers(keys %dict)) {
	print "$developer\n";
	my %seen;
	foreach my $package (sort @{$dict{$developer}}) {
		next if $seen{$package};
		$seen{$package}=1;
		print "   $package\n";
	}
	print "\n";
}

exit($errors);
